src/nvim/ex_cmds.c

```diff
@@ -6189,24 +6179,32 @@ static void incsub_display(char_u *pat, char_u *sub,
 void ex_substitute(exarg_T *eap)
 {
   if (*p_ics == NUL || !eap->is_live) {  // 'incsubstitute' is disabled
-    do_sub(eap);
+    (void)do_sub(eap);
     return;
   }
 
   char_u *save_eap = eap->arg;
   save_search_patterns();
+  int save_changedtick = curbuf->b_changedtick;
   long save_b_p_ul = curbuf->b_p_ul;
   curbuf->b_p_ul = LONG_MAX;  // make sure we can undo all changes
   block_autocmds();   // disable events before incsub opening window/buffer
   emsg_off++;         // No error messages for live commands
 
-  do_sub(eap);
+  buf_T *incsub_buf = do_sub(eap);
 
   emsg_off--;
+  if (save_changedtick != curbuf->b_changedtick
+      && !u_undo_and_forget(1)) {
```

Comment:
`u_undo_and_forget()` does not restore `b:changedtick`, which is used by some plugins like repeat.vim. @bfredl @KillTheMule should we restore it in `u_undo_and_forget()` or would it not be appropriate there?